src/plugins/loadPlugin.ts

```diff
@@ -0,0 +1,110 @@
+import * as fs from 'fs'
+import * as path from 'path'
+
+import { clearError, processError } from '../error'
+import { PluginConfig, PluginJsonConfig, PluginsServer } from '../types'
+import { getFileFromArchive } from '../utils'
+import { createLazyPluginVM } from '../vm/lazy'
+
+export async function loadPlugin(server: PluginsServer, pluginConfig: PluginConfig): Promise<boolean> {
+    const { plugin } = pluginConfig
+
+    if (!plugin) {
+        return false
+    }
+
+    try {
+        if (plugin.url?.startsWith('file:')) {
+            const pluginPath = path.resolve(server.BASE_DIR, plugin.url.substring(5))
+            const configPath = path.resolve(pluginPath, 'plugin.json')
+
+            let config: PluginJsonConfig = {}
+            if (fs.existsSync(configPath)) {
+                try {
+                    const jsonBuffer = fs.readFileSync(configPath)
+                    config = JSON.parse(jsonBuffer.toString())
+                } catch (e) {
+                    await processError(
+                        server,
+                        pluginConfig,
+                        `Could not load posthog config at "${configPath}" for plugin "${plugin.name}"`
+                    )
+                    return false
+                }
+            }
+
+            if (!config['main'] && !fs.existsSync(path.resolve(pluginPath, 'index.js'))) {
+                await processError(
+                    server,
+                    pluginConfig,
+                    `No "main" config key or "index.js" file found for plugin "${plugin.name}"`
+                )
+                return false
+            }
+
+            const jsPath = path.resolve(pluginPath, config['main'] || 'index.js')
+            const indexJs = fs.readFileSync(jsPath).toString()
+
+            const libPath = path.resolve(pluginPath, config['lib'] || 'lib.js')
+            const libJs = fs.existsSync(libPath) ? fs.readFileSync(libPath).toString() : ''
```

Comment:
One of the things you could do (though it's conceptually separate and might be better for a different PR) is to remove support for `lib.js` while at it. Removing deprecated codepaths should probably fall under "getting out of beta".